File: 18.03.2023/js/main.js
// Refactor provided algorithms with awful running time of 2^n
// (as this function iterates through ALL 2^n-1 possibilities for a log of length n).
// function cutLog(p, n) {
//     if (n == 0) return 0;
//     let q = -1;
//     for (let i = 1; i <= n; i++) {
//         q = Math.max(q, p[i] + cutLog(p, n - i));
//     }
//     return q;
// }


/* Solution one */
// time complexity of the function is O(n^2)
export const cutLog = (p, n) => {
    let dp = new Array(n + 1).fill(0);

    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= i; j++) {
            dp[i] = Math.max(dp[i], p[j] + dp[i - j]);
        }
    }

    return dp[n];
}

File: 18.03.2023/py/main.py
# Refactor provided algorithms with awful running time of 2^n
# (as this function iterates through ALL 2^n-1 possibilities for a log of length n).
# def cut_log(p, n):
#     if (n == 0):
#         return 0
#     q = -1
#     for i in range(1, n+1):
#         q = max(q, p[i] + cut_log(p, n-i))
#     return q


class WrongType(Exception):
    pass

# Solution one
# time complexity is O(n^2)
def cut_log(p, n):
    memo = [0] * (n+1)  # initialize memoization array

    for i in range(1, n+1):
        max_val = float('-inf')

        for j in range(1, i+1):
            if j == i:
                max_val = max(max_val, p[j])
            else:
                max_val = max(max_val, memo[j] + memo[i-j])

        memo[i] = max_val

    return memo[n]


import cProfile
import time
from timeit import default_timer as timer


if __name__ == '__main__':
    cProfile.run('cut_log(230, 4, xs)')

    start_time = time.time()
    cut_log(230, 4, xs)
    end_time = time.time()

    print('', end_time - start_time)
